File: frontend/src/api/gate.ts
import { http } from "@/utils/http";

export type Decision = "Allow" | "Block" | "Degrade" | "Deny" | "HumanApproval";

export type GateDecision = {
  request_id: string;
  timestamp: string;
  gate_type: "message" | "action" | "return";
  decision: Decision;
  risk_score: number;
  risk_level: "low" | "medium" | "high" | "critical";
  matched_rules: string[];
  reason: string;
  tool_name?: string;
  agent_id?: string;
  token_status?: string;
  auth_mode?: "strict" | "compat" | "warn";
  unauthorized_allow?: boolean;
};

type RawGateSummary = Record<string, unknown>;

type RawGateOverview = {
  message_gate: RawGateSummary;
  action_gate: RawGateSummary;
  return_gate: RawGateSummary;
  recent_decisions: GateDecision[];
};

export type GateOverview = {
  message_gate: GateSummary;
  action_gate: GateSummary;
  return_gate: GateSummary;
  recent_decisions: GateDecision[];
};

export type GateSummary = {
  status: string;
  today_count: number;
  block_count: number;
  decision_counts: Record<string, number>;
};

export type GateEvaluateRequest =
  | {
      type: "message" | "return";
      agent_id?: string;
      body?: Record<string, unknown>;
      content?: string;
    }
  | {
      type: "action";
      agent_id?: string;
      tool_name: string;
      params?: Record<string, unknown>;
      headers?: Record<string, string>;
    };

type ApiResult<T> = { success: boolean; data: T };

export function normalizeRiskScore(score: number) {
  if (!Number.isFinite(score)) return 0;
  if (score > 0 && score <= 1) return Math.round(score * 100);
  return Math.max(0, Math.min(100, Math.round(score)));
}

export function normalizeGateDecision(decision: GateDecision): GateDecision {
  return {
    ...decision,
    risk_score: normalizeRiskScore(decision.risk_score)
  };
}

export function normalizeGateOverview(raw: RawGateOverview): GateOverview {
  const readNumber = (value: unknown) => (typeof value === "number" && Number.isFinite(value) ? value : undefined);

  const buildSummary = (summary: RawGateSummary): GateSummary => {
    const status = typeof summary?.status === "string" ? summary.status : "online";
    const todayCount =
      readNumber(summary?.today_count) !== undefined
        ? (readNumber(summary?.today_count) as number)
        : Object.values(summary || {}).reduce<number>((sum, value) => {
            const numeric = readNumber(value);
            return numeric !== undefined ? sum + numeric : sum;
          }, 0);
    const blockCount =
      readNumber(summary?.block_count) !== undefined
        ? (readNumber(summary?.block_count) as number)
        : (readNumber(summary?.Block) || 0) + (readNumber(summary?.Deny) || 0);
    const decisionCounts =
      summary?.decision_counts && typeof summary.decision_counts === "object"
        ? (summary.decision_counts as Record<string, number>)
        : (summary as Record<string, number>);

    return {
      status,
      today_count: todayCount,
      block_count: blockCount,
      decision_counts: decisionCounts || {}
    };
  };

  return {
    message_gate: buildSummary(raw.message_gate),
    action_gate: buildSummary(raw.action_gate),
    return_gate: buildSummary(raw.return_gate),
    recent_decisions: (raw.recent_decisions || []).map(normalizeGateDecision)
  };
}

export const getGateOverview = () => {
  return http.request<ApiResult<RawGateOverview>>("get", "/aegis/gate/overview");
};

export const getGateDecisions = (params?: object) => {
  return http.request<ApiResult<GateDecision[]>>("get", "/aegis/gate/decisions", {
    params
  });
};

export const evaluateGate = (data?: GateEvaluateRequest) => {
  return http.request<ApiResult<GateDecision>>("post", "/aegis/gate/evaluate", {
    data
  });
};

File: backend/internal/contract/gate.go
package contract

import (
	"net/http"

	"aegisguard/internal/interfaces"
)

// ============================================================================
// [分工1+3] 执行平面 → 控制平面：门控评估与查询
// ============================================================================

// GateQuery [UNDEFINED] [分工1+3] 三门决策查询接口（只读）
//
// 执行平面暴露给控制平面的能力：查询三门概览统计和决策历史。
type GateQuery interface {
	Overview() (*GateOverview, error)
	Decisions(limit int, gateType, action string) ([]interfaces.GateDecision, error)
}

// GateEvaluator [UNDEFINED] [分工1+3] 手动触发三门评估接口（调试/模拟用）
//
// 执行平面暴露给控制平面的能力：手动触发门控评估，返回决策。
// 前端对应：gate.ts evaluateGate() -> POST /aegis/gate/evaluate
type GateEvaluator interface {
	EvaluateMessage(requestID string, body []byte, agentID string) interfaces.EvaluateResult
	EvaluateAction(requestID string, toolName string, params map[string]interface{}, headers http.Header, agentID string) interfaces.EvaluateResult
	EvaluateReturn(requestID string, body []byte, agentID string) interfaces.EvaluateResult
}

// GateOverview [分工3] 三门概览统计数据
type GateOverview struct {
	MessageGate     map[string]int            `json:"message_gate"`
	ActionGate      map[string]int            `json:"action_gate"`
	ReturnGate      map[string]int            `json:"return_gate"`
	RecentDecisions []interfaces.GateDecision `json:"recent_decisions"`
}
